src/DataStructures/Tensor/EagerMath/DotProduct.hpp

```diff
@@ -0,0 +1,60 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines functions euclidean dot_product and dot_product with a metric
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute the Euclidean dot product of two rank-1 tensors
+ *
+ * \details
+ * Returns \f$A^a B^b \delta_{ab}\f$ for input tensors \f$A^a\f$ and \f$B^b\f$.
+ */
+template <typename DataType, typename Index>
+Scalar<DataType> dot_product(
+    const Tensor<DataType, Symmetry<1>, tmpl::list<Index>>& vector_a,
+    const Tensor<DataType, Symmetry<1>, tmpl::list<Index>>& vector_b) noexcept {
+  auto dot_product = make_with_value<Scalar<DataType>>(vector_a, 0.);
+  for (size_t d = 0; d < Index::dim; ++d) {
+    get(dot_product) += vector_a.get(d) * vector_b.get(d);
+  }
+  return dot_product;
+}
+
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute the dot_product of a rank-1 tensor
+ *
+ * \details
+ * Returns \f$g_{ab} A^a B^b\f$, where \f$g_{ab}\f$ is the metric,
+ * \f$A^a\f$ is vector_a, and \f$B^b\f$ is vector_b.
+ */
+template <typename DataType, typename Index0, typename Index1, typename Index2>
```

Comment:
there should only be one `typename Index`, as both vectors need to be the same, and the metric should have the opposite valence which is given with`change_index_up_lo<Index>`  (then the static_asserts are not needed)